# C1_성민지_Python을 이용한 실습_220804

---



In [51]:
import pandas as pd
import numpy as np
from scipy import stats
from scipy.stats import norm

### Python을 이용한 기술통계 실습 (210쪽)
- 자동차 연비 Data Set에서 기술통계치 구하기

In [52]:
dat = pd.read_csv('/home/piai/minji/dataset/mycars.csv', encoding = 'euc-kr')

print(dat.shape)
dat.head()

(234, 11)


manufacturer model  displacement  year  cylinder automatic driving  mpg  \
0         audi    a4           1.8  1999         4      auto       f   18   
1         audi    a4           1.8  1999         4    manual       f   21   
2         audi    a4           2.0  2008         4    manual       f   20   
3         audi    a4           2.0  2008         4      auto       f   21   
4         audi    a4           2.8  1999         6      auto       f   16   

   highway_mileage fuel    class  
0               29    p  compact  
1               29    p  compact  
2               31    p  compact  
3               30    p  compact  
4               26    p  compact

In [53]:
dat.groupby(['automatic', 'mpg']).describe().round(3)

displacement                                              year  \
                     count   mean    std  min    25%   50%    75%  max count   
automatic mpg                                                                  
auto      9            4.0  4.700  0.000  4.7  4.700  4.70  4.700  4.7   4.0   
          11          18.0  5.144  0.694  3.3  4.825  5.30  5.400  6.1  18.0   
          12           6.0  5.167  0.709  4.2  4.650  5.40  5.550  6.0   6.0   
          13          20.0  4.845  0.630  3.9  4.600  4.70  5.475  5.7  20.0   
          14          16.0  4.569  0.784  3.3  4.000  4.70  4.850  6.5  16.0   
          15          12.0  4.067  1.052  2.8  3.400  3.80  4.600  6.2  12.0   
          16          16.0  3.456  0.841  1.8  2.800  3.55  4.000  5.3  16.0   
          17          10.0  3.250  0.337  2.7  3.025  3.20  3.525  3.8  10.0   
          18          17.0  2.859  0.595  1.8  2.500  3.00  3.100  3.8  17.0   
          19          13.0  2.546  0.654  2.0  2.000  2.40  3.300  3.5  13.0   
          20           5.0  2.400  0.224  2.0  2.500  2.50  2.500  2.5   5.0   
          21           7.0  2.271  0.170  2.0  2.200  2.20  2.400  2.5   7.0   
          22           4.0  2.200  0.231  2.0  2.000  2.20  2.400  2.4   4.0   
          23           1.0  2.500    NaN  2.5  2.500  2.50  2.500  2.5   1.0   
          24           5.0  1.720  0.110  1.6  1.600  1.80  1.800  1.8   5.0   
          25           1.0  1.800    NaN  1.8  1.800  1.80  1.800  1.8   1.0   
          26           1.0  1.800    NaN  1.8  1.800  1.80  1.800  1.8   1.0   
          29           1.0  1.900    NaN  1.9  1.900  1.90  1.900  1.9   1.0   
manual    9            1.0  4.700    NaN  4.7  4.700  4.70  4.700  4.7   1.0   
          11           2.0  5.200  0.000  5.2  5.200  5.20  5.200  5.2   2.0   
          12           2.0  4.700  0.000  4.7  4.700  4.70  4.700  4.7   2.0   
          13           1.0  4.600    NaN  4.6  4.600  4.60  4.600  4.6   1.0   
          14           3.0  4.500  0.794  3.9  4.050  4.20  4.800  5.4   3.0   
          15          12.0  3.875  1.169  2.7  3.250  3.55  4.150  7.0  12.0   
          16           3.0  4.867  1.893  2.7  4.200  5.70  5.950  6.2   3.0   
          17           6.0  2.967  0.509  2.7  2.725  2.80  2.800  4.0   6.0   
          18           9.0  2.733  0.545  1.8  2.500  2.80  3.000  3.8   9.0   
          19           7.0  2.600  0.503  2.0  2.350  2.50  2.750  3.5   7.0   
          20           6.0  2.250  0.274  2.0  2.000  2.25  2.500  2.5   6.0   
          21          16.0  2.131  0.227  1.8  2.000  2.00  2.400  2.5  16.0   
          23           2.0  2.050  0.636  1.6  1.825  2.05  2.275  2.5   2.0   
          25           1.0  1.600    NaN  1.6  1.600  1.60  1.600  1.6   1.0   
          26           2.0  1.800  0.000  1.8  1.800  1.80  1.800  1.8   2.0   
          28           2.0  1.700  0.141  1.6  1.650  1.70  1.750  1.8   2.0   
          33           1.0  1.900    NaN  1.9  1.900  1.90  1.900  1.9   1.0   
          35           1.0  1.900    NaN  1.9  1.900  1.90  1.900  1.9   1.0   

                         ... cylinder      highway_mileage                 \
                   mean  ...      75%  max           count    mean    std   
automatic mpg            ...                                                
auto      9    2008.000  ...      8.0  8.0             4.0  12.000  0.000   
          11   2001.000  ...      8.0  8.0            18.0  15.444  0.984   
          12   2008.000  ...      8.0  8.0             6.0  17.833  0.408   
          13   2005.300  ...      8.0  8.0            20.0  17.500  0.889   
          14   2004.062  ...      8.0  8.0            16.0  18.125  1.258   
          15   2001.250  ...      8.0  8.0            12.0  21.667  2.229   
          16   2002.938  ...      6.0  8.0            16.0  23.000  2.221   
          17   2006.200  ...      6.0  6.0            10.0  24.700  1.418   
          18   2001.647  ...      6.0  6.0    

- 전체 데이터 개수는 234개이다.
- 년도는 1999년부터 2008년까지 분포되어 있는 것으로 확인된다.
- 차에 배치된 실린더의 개수는 일반적으로 5-6개이다.

In [54]:
dat.groupby(['automatic', 'highway_mileage']).describe().round(3)

displacement                                         \
                                 count   mean    std  min    25%   50%    75%   
automatic highway_mileage                                                       
auto      12                       4.0  4.700  0.000  4.7  4.700  4.70  4.700   
          14                       2.0  5.700  0.566  5.3  5.500  5.70  5.900   
          15                      10.0  5.190  0.608  4.0  4.825  5.25  5.625   
          16                       3.0  5.067  0.416  4.6  4.900  5.20  5.300   
          17                      25.0  4.724  0.804  3.3  4.000  4.70  5.400   
          18                       9.0  5.089  0.772  3.7  4.400  5.40  5.700   
          19                      11.0  4.282  0.618  3.4  3.850  4.60  4.700   
          20                       8.0  4.000  0.983  2.7  3.675  4.00  4.325   
          21                       2.0  4.200  0.566  3.8  4.000  4.20  4.400   
          22                       5.0  3.600  0.628  3.0  3.300  3.30  3.800   
          23                       6.0  3.833  1.140  2.5  3.050  3.90  4.150   
          24                       8.0  3.000  0.524  2.4  2.650  2.90  3.300   
          25                       9.0  3.567  1.382  1.8  2.800  3.10  3.800   
          26                      20.0  2.770  0.580  2.0  2.350  2.80  3.100   
          27                      10.0  2.590  0.606  2.0  2.200  2.40  2.950   
          28                       4.0  3.150  0.794  2.0  2.975  3.40  3.575   
          29                       7.0  2.300  0.606  1.8  1.900  2.00  2.500   
          30                       4.0  2.150  0.300  1.8  1.950  2.20  2.400   
          31                       3.0  2.433  0.058  2.4  2.400  2.40  2.450   
          32                       2.0  1.600  0.000  1.6  1.600  1.60  1.600   
          33                       1.0  1.800    NaN  1.8  1.800  1.80  1.800   
          35                       1.0  1.800    NaN  1.8  1.800  1.80  1.800   
          36                       2.0  1.800  0.000  1.8  1.800  1.80  1.800   
          41                       1.0  1.900    NaN  1.9  1.900  1.90  1.900   
manual    12                       1.0  4.700    NaN  4.7  4.700  4.70  4.700   
          16                       4.0  4.800  0.271  4.6  4.675  4.70  4.825   
          17                       6.0  3.900  0.727  3.3  3.400  3.65  4.125   
          18                       1.0  4.000    NaN  4.0  4.000  4.00  4.000   
          19                       2.0  3.850  0.212  3.7  3.775  3.85  3.925   
          20                       3.0  3.600  1.559  2.7  2.700  2.70  4.050   
          22                       2.0  3.650  1.344  2.7  3.175  3.65  4.125   
          23                       1.0  4.600    NaN  4.6  4.600  4.60  4.600   
          24                       5.0  3.600  1.901  2.7  2.700  2.80  2.800   
          25                       6.0  2.733  0.273  2.5  2.500  2.65  2.950   
          26                      12.0  3.358  1.358  1.8  2.500  2.90  3.850   
          27                       4.0  2.725  0.519  2.4  2.475  2.50  2.750   
          28                       3.0  2.167  0.289  2.0  2.000  2.00  2.250   
          29                      15.0  2.033  0.226  1.6  2.000  2.00  2.100   
          31                       4.0  2.300  0.200  2.0  2.300  2.40  2.400   
          32                       2.0  2.050  0.636  1.6  1.825  2.05  2.275   
          33                       1.0  1.600    NaN  1.6  1.600  1.60  1.600   
          34                       1.0  1.800    NaN  1.8  1.800  1.80  1.800   
          35                       1.0  1.800    NaN  1.8  1.800  1.80  1.800   
          37                       1.0  1.800    NaN  1.8  1.800  1.80  1.800   
          44                       2.0  1.900  0.000  1.9  1.900  1.90  1.900   

                                year            ... cylinder        mpg  \
                           max count      mean  ...      75%  

### Python을 이용한 기술통계 연습 (212쪽)

In [55]:
dat = pd.read_csv('/home/piai/minji/dataset/mycars.csv', encoding = 'euc-kr')

print(dat.shape)
dat.head()

(234, 11)


manufacturer model  displacement  year  cylinder automatic driving  mpg  \
0         audi    a4           1.8  1999         4      auto       f   18   
1         audi    a4           1.8  1999         4    manual       f   21   
2         audi    a4           2.0  2008         4    manual       f   20   
3         audi    a4           2.0  2008         4      auto       f   21   
4         audi    a4           2.8  1999         6      auto       f   16   

   highway_mileage fuel    class  
0               29    p  compact  
1               29    p  compact  
2               31    p  compact  
3               30    p  compact  
4               26    p  compact

In [56]:
dat1 = dat[['automatic', 'model']]

print(dat1.shape)
dat1.head()

(234, 2)


automatic model
0      auto    a4
1    manual    a4
2    manual    a4
3      auto    a4
4      auto    a4

In [57]:
# 'automatic' 변수의 범주별 개수
count_auto = dat1['automatic'].value_counts().sort_index()


# 누적개수, 확률, 누적확률
cumcnt_auto = np.cumsum(count_auto)
percent_auto = count_auto / sum(count_auto) * 100
cumpct_auto = np.cumsum(percent_auto)

In [58]:
count_data_auto = pd.DataFrame({'Count' : count_auto, 'CumCnt' : cumcnt_auto,
                                'Percent' : percent_auto, 'CumPct' : cumpct_auto})

In [59]:
count_data_auto

Count  CumCnt    Percent      CumPct
auto      157     157  67.094017   67.094017
manual     77     234  32.905983  100.000000

**구동방식을 기준으로 데이터를 구분**
- 자동 157개와 수동 77개로 이루어져 있다는 것을 확인할 수 있다.
- manual의 CumCnt를 통해 전체 데이터의 개수가 234개인 것을 확인할 수 있다.
- auto는 전체 데이터의 약 67%를 차지한다.

### Python을 이용한 기술통계 연습

In [19]:
# 'model' 변수의 범주별 개수
count_model = dat1['model'].value_counts().sort_index()


# 누적개수, 확률, 누적확률
cumcnt_model = np.cumsum(count_model)
percent_model = count_model / sum(count_model) * 100
cumpct_model = np.cumsum(percent_model)

In [20]:
count_data_model = pd.DataFrame({'Count' : count_model, 'CumCnt' : cumcnt_model,
                                'Percent' : percent_model, 'CumPct' : cumpct_model})

In [21]:
count_data_model

Count  CumCnt   Percent      CumPct
4runner 4wd                 6       6  2.564103    2.564103
a4                          7      13  2.991453    5.555556
a4 quattro                  8      21  3.418803    8.974359
a6 quattro                  3      24  1.282051   10.256410
altima                      6      30  2.564103   12.820513
c1500 suburban 2wd          5      35  2.136752   14.957265
camry                       7      42  2.991453   17.948718
camry solara                7      49  2.991453   20.940171
caravan 2wd                11      60  4.700855   25.641026
civic                       9      69  3.846154   29.487179
corolla                     5      74  2.136752   31.623932
corvette                    5      79  2.136752   33.760684
dakota pickup 4wd           9      88  3.846154   37.606838
durango 4wd                 7      95  2.991453   40.598291
expedition 2wd              3      98  1.282051   41.880342
explorer 4wd                6     104  2.564103   44.444444
f150 pickup 4wd             7     111  2.991453   47.435897
forester awd                6     117  2.564103   50.000000
grand cherokee 4wd          8     125  3.418803   53.418803
grand prix                  5     130  2.136752   55.555556
gti                         5     135  2.136752   57.692308
impreza awd                 8     143  3.418803   61.111111
jetta                       9     152  3.846154   64.957265
k1500 tahoe 4wd             4     156  1.709402   66.666667
land cruiser wagon 4wd      2     158  0.854701   67.521368
malibu                      5     163  2.136752   69.658120
maxima                      3     166  1.282051   70.940171
mountaineer 4wd             4     170  1.709402   72.649573
mustang                     9     179  3.846154   76.495726
navigator 2wd               3     182  1.282051   77.777778
new beetle                  6     188  2.564103   80.341880
passat                      7     195  2.991453   83.333333
pathfinder 4wd              4     199  1.709402   85.042735
ram 1500 pickup 4wd        10     209  4.273504   89.316239
range rover                 4     213  1.709402   91.025641
sonata                      7     220  2.991453   94.017094
tiburon                     7     227  2.991453   97.008547
toyota tacoma 4wd           7     234  2.991453  100.000000

**모델을 기준으로 데이터를 구분**
- caravan 2wd과 ram 1500 pickup 4wd이 여러 자동차 모델 중 가장 많이 분포되어 있는 것으로 확인된다.

### Python을 이용한 확률 계산 실습 (226쪽)

In [38]:
mu = 175
sigma = 5
x = 180

prob = stats.norm.cdf(x, mu, sigma)
print("나의 키가 {0}cm일 경우 내가 위치할 % 범위 : {1:.1f}%".format(x, (1-prob) * 100))

나의 키가 180cm일 경우 내가 위치할 % 범위 : 15.9%


### Python을 이용한 확률 계산 실습 (228쪽)

In [47]:
# 1)

mu = 12
sigma = 3
x1 = 15

prob1 = stats.norm.cdf(15, 12, 3) - stats.norm.cdf(9, 12, 3)
print("{0}km 이상 {1}km 이하를 달릴 확률은 {2:.1f}%에 해당한다.".format(9, 15, prob1 * 100))

9km 이상 15km 이하를 달릴 확률은 68.3%에 해당함


In [49]:
# 2)

prob2 = stats.norm.cdf(15, 12, 3)
print("{0}km 이상 {1}km 이하를 달릴 확률은 {2:.1f}%에 해당한다.".format(12, 15, (1 - prob2) * 100))

12km 이상 15km 이하를 달릴 확률은 15.9%에 해당한다.


### Python을 이용한 확률 계산 실습 (236쪽)

In [30]:
t = 2.0
df = 100

prob = stats.t.cdf(t, df)
print("P(T <= F) : {:.3f}".format(prob))

P(T <= F) : 0.976


자유도가 100일 때, 2.0에 대한 검정통계량 T의 누적확률은 0.976이다.

### Python을 이용한 확률 계산 실습 (239쪽)

In [32]:
chisq = 3
df = 8

prob = stats.chi2.cdf(chisq, df)
print("P(X <= {0}) : {1:.4f}".format(chisq, prob))

P(X <= 3) : 0.0656


자유도가 8일 때, 3에 대한 검정통계량 X^2의 누적확률은 0.0656이다.

### Python을 이용한 확률분포 그리기, 확률 계산 실습 (243쪽)

In [34]:
f = 1.5
dfnum = 10
dfden = 100

prob = stats.f.cdf(x = f, dfn = dfnum, dfd = dfden)
print("P(X <= F): {:.3f}".format(prob))

P(X <= F): 0.850


자유도가 각각 10과 100일 때, F = 1.5에 대한 검정통계량 F의 누적확률은 0.850이다.

### Python을 이용한 확률분포 그리기, 확률 계산 실습 (250쪽)

In [36]:
n = 5
for i in range(n+1) :
    prob = stats.binom.pmf(k = i, n = n, p = 0.3)
    print("P(X={0}) = {1:.3f}".format(i, prob))

P(X=0) = 0.168
P(X=1) = 0.360
P(X=2) = 0.309
P(X=3) = 0.132
P(X=4) = 0.028
P(X=5) = 0.002


5대의 차량을 임의로 선택했을 때, 불량대수가 각각 0, 1, 2, 3, 4, 5일 확률은 각각 0.168, 0.360, 0.309, 0.132, 0.028, 0.002이다.